src/isotools/_utils.py

```diff
@@ -561,3 +561,135 @@ def cmp_dist(a, b, min_dist=3):
+
+
+# gene structure variation
+
+def structure_feature_cov(transcripts, samples, feature='TSS'):
+    '''
+    :param transcripts: A list of transcript annotations of a gene obtained from isoseq[gene].transcripts.
+    :param feature: 'ec', 'TSS', 'PAS'.
+    :param samples: A list of sample names to specify the samples to be considered.
+    :return: Selected feature of input transcripts and corresponding coverage across samples.
+
+    1) ec - exon_chain, query coverage matrix and exon positions, and return all the exon_chain and coverage.
+    2) TSS/PAS, query TSS_unified or PAS_unified coverage matrix, and return all the positions and coverage.
+    '''
+
+    if feature not in ['ec', 'TSS', 'PAS']:
+        raise ValueError('choose feature from ec, TSS, PAS')
+
+    cov = {}
+    if feature == 'ec':
+        field = 'coverage'
+        for tr in transcripts:
+            if tr[field] is None:
+                continue
+
+            # Convert list of exons to tuple of tuples to make it hashable as a key of a dictionary
+            exon_chain = tuple(map(tuple, tr['exons']))
+            if exon_chain not in cov:
+                cov[exon_chain] = 0
+            for s, n in tr[field].items():
+                if s in samples:
+                    cov[exon_chain] += n
+            if cov[exon_chain] == 0:
+                del cov[exon_chain]
+    else:
+        field = f'{feature}_unified'
+        for tr in transcripts:
+            if tr[field] is None:
+                continue
+
+            for s,pos_dict in tr[field].items():
+                if s in samples:
+                    for pos, n in pos_dict.items():
+                        if pos not in cov:
+                            cov[pos] = 0
+                        cov[pos] += n
+
+    # keep ones with coverage > 0
+    cov = {k: v for k, v in cov.items() if v > 0}
+
+    if len(cov) == 0:
+        return [], []
+
+    # sort the coverage in descending order and return
+    abundance, occurrence = zip(*sorted(zip(list(cov.values()), list(cov.keys())), reverse=True))
+    return list(abundance), list(occurrence)
+
+
+def count_distinct_pos(pos_list, strict_pos=15):
+    '''
+    :param pos_list: A list of TSS/PAS positions, sorted by their abundance descendingly (output from structure_feature_cov).
+    :param strict_pos: Difference allowed between two positions when considering identical TSS/PAS.
+    :return: How many distinct positions are there.
+    '''
+
+    merged_idx = []
+    for x in range(len(pos_list)-1):
+        if x in merged_idx:
+            continue
+        for y in range(x+1, len(pos_list)):
+            if y in merged_idx:
+                continue
+            if pos_list[x] - strict_pos <= pos_list[y] <= pos_list[x] + strict_pos:
+                #print(f'{x}:{pos_list[x]} and {y}: {pos_list[y]} are close enough')
+                # keep the one with higher coverage
+                merged_idx.append(y)
+    return len(pos_list) - len(merged_idx)
```

Comment:
```py
sorted_pos = sorted(pos_list)
return sum([1 for i in pos_list[:-1] if pos_list[i+1] - i >= strict_pos])
```
Isn't this the same? Or is it important that high abundance positions are added first?

Reply:
Not really sure what you're doing here.